Common storage backend docs (#2347)
grafana/mimir: docs/sources/operators-guide/configure/configure-object-storage-backend.md

---
title: "Configure Grafana Mimir object storage backend"
menuTitle: "Configure object storage"
description: "Learn how to configure Grafana Mimir to use different object storage backend implementations."
weight: 70
---

# Configure Grafana Mimir object storage backend

Grafana Mimir can use different object storage services to persist blocks containing the metrics data, as well as recording rules and alertmanager state.
The supported backends are:

- [Amazon S3](https://aws.amazon.com/s3/) (and compatible implementations like [MinIO](https://min.io/))
- [Google Cloud Storage](https://cloud.google.com/storage)
- [Azure Blob Storage](https://azure.microsoft.com/es-es/services/storage/blobs/)
- [Swift (OpenStack Object Storage)](https://wiki.openstack.org/wiki/Swift)

> Additionally, a file system emulated [`filesystem`]({{< relref "../configure/reference-configuration-parameters/index.md#filesystem_storage_backend" >}}) object storage implementation can be also used for testing purposes (it's not recommended for production workloads).

Reviewer: ```suggestion
Additionally and for non-production testing purposes, you can use a file-system emulated [`filesystem`]({{< relref "../configure/reference-configuration-parameters/index.md#filesystem_storage_backend" >}}) object storage implementation.
```

> [Ruler and alertmanager support a `local` implementation]({{< relref "../architecture/components/ruler/index.md#local-storage" >}}), which is similar to `filesystem` in the way that it uses the local file system, but it is a read-only data source and can be used to provision state into those components.

Reviewer: ```suggestion
[Ruler and alertmanager support a `local` implementation]({{< relref "../architecture/components/ruler/index.md#local-storage" >}}),
which is similar to `filesystem` in the way that it uses the local file system,
but it is a read-only data source and can be used to provision state into those components.
```
Semantic line breaks; no change to content.

## Common configuration

> The [common configuration]({{< relref "about-configurations.md#common-configurations" >}}) can be used to avoid repetition by filling the [`common`]({{< relref "../configure/reference-configuration-parameters/index.md#common" >}}) configuration block or by providing the `-common.storage.*` CLI flags.

Reviewer: ```suggestion
To avoid repetition, you can use the [common configuration]({{< relref "about-configurations.md#common-configurations" >}}) and fill the [`common`]({{< relref "../configure/reference-configuration-parameters/index.md#common" >}}) configuration block or by providing the `-common.storage.*` CLI flags.
```
Use semantic line breaks.

Reviewer: Reworded to change passive voice to active voice.

> Note that blocks storage can't be located in the same path of the same bucket as the ruler and alertmanager stores, so when using the common configuration, [`blocks_storage`]({{< relref "../configure/reference-configuration-parameters/index.md#blocks_storage" >}}) should either:

Reviewer: ```suggestion
> **Note:** Blocks storage cannot be located in the same path of the same bucket as the ruler and alertmanager stores. When using the common configuration, make [`blocks_storage`]({{< relref "../configure/reference-configuration-parameters/index.md#blocks_storage" >}}) use either a:
```

Reviewer: Use semantic line breaks.

> - use a different bucket, overriding the common bucket name
> - use a storage prefix

Reviewer: ```suggestion
- different bucket, overriding the common bucket name
- storage prefix
```

> Grafana Mimir will fail to start if blocks storage is configured to use the same bucket and storage prefix as alertmanager or ruler store.

Reviewer: ```suggestion
Grafana Mimir will fail to start if you configure blocks storage to use the same bucket and storage prefix that the  alertmanager or ruler store uses.
```

Reviewer: Vet for technical accuracy.

Author: It's correct.

> A valid configuration for the object storage (taken from the ["Play with Grafana Mimir" tutorial](https://grafana.com/tutorials/play-with-grafana-mimir/)) would be:

Reviewer: ```suggestion
A valid configuration for the object storage (taken from the ["Play with Grafana Mimir" tutorial](https://grafana.com/tutorials/play-with-grafana-mimir/)) looks as follows:
```

```yaml
common:
  storage:
    backend: s3
    s3:
      endpoint: minio:9000
      access_key_id: mimir
      secret_access_key: supersecret
      insecure: true
      bucket_name: mimir

blocks_storage:
  storage_prefix: blocks
```

> > **Note**: If you're using a mix of YAML files and CLI flags, pay attention to the [precedence logic]({{< relref "about-configurations.md#common-configurations" >}}) of those.

Reviewer: ```suggestion
> **Note**: If you're using a mixture of YAML files and CLI flags, pay attention to their [precedence logic]({{< relref "about-configurations.md#common-configurations" >}}).
```
